src/garbage_collector.rs

```diff
@@ -164,33 +187,92 @@ impl GarbageCollectorOrchestrator {
         Ok(())
     }
 
-    /// Collect garbage from the WAL SSTs. This will delete any WAL SSTs that are
-    /// older than the minimum age specified in the options and are also older than
-    /// the last compacted WAL SST.
-    async fn collect_garbage_wal_ssts(&self) -> Result<(), SlateDBError> {
-        let utc_now = Utc::now();
-        let last_compacted_wal_sst_id = self
-            .manifest_store
-            // Get the latest manifest so we can get the last compacted id
-            .read_latest_manifest()
+    async fn load_stored_manifest(&self) -> Result<StoredManifest, SlateDBError> {
+        StoredManifest::load(Arc::clone(&self.manifest_store))
             .await?
-            .ok_or_else(|| SlateDBError::ManifestMissing)?
-            // read_latest_manifest returns (id, manifest) but we only care about the manifest
-            .1
-            .core
-            .last_compacted_wal_sst_id;
+            .ok_or_else(|| SlateDBError::ManifestMissing)
+    }
+
+    fn filter_expired_checkpoints(
+        manifest: &StoredManifest,
+    ) -> Result<Option<CoreDbState>, SlateDBError> {
+        let utc_now = Utc::now();
+        let mut db_state = manifest.db_state().clone();
+        let retained_checkpoints: Vec<Checkpoint> = db_state
+            .checkpoints
+            .iter()
+            .filter(|checkpoint| match checkpoint.expire_time {
+                Some(expire_time) => DateTime::<Utc>::from(expire_time) > utc_now,
+                None => true,
+            })
```

Comment:
I'm not certain I understand the checkpointing RFC well enough to comment meaningfully, but is there a potential race condition if clients refresh the expire time?:

```
   // The UTC unix timestamp seconds that a checkpoint expires at. Clients may update this value.
   // If `checkpoint_expire_time_s` is older than now(), the checkpoint is considered expired.
   // If `checkpoint_expire_time_s` is 0, the checkpoint will never expire.
   checkpoint_expire_time_s: uint;
```

What if a client refreshes the timestamp while this iteration is going on? I think having #356 might help this situation since at least the timestamp of the GC will always be no later than the latest operation that was persisted to the manifest (and then we can guarantee that a refresh operation fails if its trying to refresh an already-expired checkpoint).

I think almost everything else in S3 is effectively immutable, so I imagine we don't have a problem like this elsewhere.

Reply:
Makes sense to me, I missed that part of the RFC!

Reply:
Event without the heuristic to update expiration time when half of its age has passed, doesn't the conflict on writing the manifest guarantee this? If the expiration time changed after the manifest was read, manifest write from this method would fail, and `filter_expired_checkpoints` would be rerun. Rerun would use the new expiration.

Reply:
> Event without the heuristic to update expiration time when half of its age has passed, doesn't the conflict on writing the manifest guarantee this? If the expiration time changed after the manifest was read, manifest write from this method would fail, and filter_expired_checkpoints would be rerun. Rerun would use the new expiration.

Yep, this is the way it's meant to work. Either the GC wins and the process that's refreshing it's checkpoint fails, or the process refreshing it's checkpoint wins and the GC will fail to remove the checkpoint. It's not possible for the process refreshing the checkpoint to successfully refresh it, only to have the GC then go an remove it.